File: modules/StudentManager.js
const readlineSync = require('readline-sync');

class StudentManager {
    constructor() {
        this.students = [];
        this.nextId = 1;
    }

    register() {
        console.log('\n--- ثبت نام دانشجو ---');
        const username = readlineSync.question('نام کاربری: ');
        
        // Check if username already exists
        if (this.students.find(s => s.username === username)) {
            console.log('این نام کاربری قبلاً استفاده شده است!');
            return null;
        }

        const password = readlineSync.question('رمز عبور: ', { hideEchoBack: true });
        const confirmPassword = readlineSync.question('تکرار رمز عبور: ', { hideEchoBack: true });

        if (password !== confirmPassword) {
            console.log('رمزهای عبور مطابقت ندارند!');
            return null;
        }

        const student = {
            id: this.nextId++,
            username: username,
            password: password,
            isActive: true,
            registrationDate: new Date()
        };

        this.students.push(student);
        console.log('ثبت نام با موفقیت انجام شد!');
        return student;
    }

    login(username, password) {
        const student = this.students.find(s => s.username === username && s.password === password && s.isActive);
        return student || null;
    }

    searchBooks(bookManager) {
        console.log('\n--- جستجوی کتاب ---');
        console.log('1. جستجو بر اساس عنوان');
        console.log('2. جستجو بر اساس سال نشر');
        console.log('3. جستجو بر اساس نام نویسنده');
        console.log('4. جستجوی ترکیبی');

        const choice = readlineSync.question('نوع جستجو را انتخاب کنید: ');

        let searchTerm;
        let results = [];

        switch (choice) {
            case '1':
                searchTerm = readlineSync.question('عنوان کتاب: ');
                results = bookManager.searchByTitle(searchTerm);
                break;
            case '2':
                searchTerm = readlineSync.question('سال نشر: ');
                results = bookManager.searchByYear(searchTerm);
                break;
            case '3':
                searchTerm = readlineSync.question('نام نویسنده: ');
                results = bookManager.searchByAuthor(searchTerm);
                break;
            case '4':
                const title = readlineSync.question('عنوان کتاب (اختیاری): ');
                const year = readlineSync.question('سال نشر (اختیاری): ');
                const author = readlineSync.question('نام نویسنده (اختیاری): ');
                results = bookManager.combinedSearch(title, year, author);
                break;
            default:
                console.log('انتخاب نامعتبر!');
                return;
        }

        if (results.length === 0) {
            console.log('هیچ کتابی یافت نشد!');
        } else {
            console.log(`\nنتایج جستجو (${results.length} کتاب):`);
            results.forEach((book, index) => {
                console.log(`${index + 1}. ${book.title} - ${book.author} (${book.publicationYear})`);
                console.log(`   وضعیت: ${book.isAvailable ? 'موجود' : 'امانت داده شده'}`);
                console.log(`   شناسه: ${book.id}`);
                console.log('');
            });
        }
    }

    requestBorrow(bookManager, borrowManager) {
        console.log('\n--- ثبت درخواست امانت ---');
        const bookId = readlineSync.question('شناسه کتاب: ');
        
        const book = bookManager.getBookById(parseInt(bookId));
        if (!book) {
            console.log('کتاب مورد نظر یافت نشد!');
            return;
        }

        if (!book.isAvailable) {
            console.log('این کتاب در حال حاضر در امانت است!');
            return;
        }

        const startDate = readlineSync.question('تاریخ شروع امانت (YYYY-MM-DD): ');
        const endDate = readlineSync.question('تاریخ پایان امانت (YYYY-MM-DD): ');

        // Validate dates
        const start = new Date(startDate);
        const end = new Date(endDate);
        const today = new Date();
        today.setHours(0, 0, 0, 0);

        if (start < today) {
            console.log('تاریخ شروع نمی‌تواند قبل از امروز باشد!');
            return;
        }

        if (end <= start) {
            console.log('تاریخ پایان باید بعد از تاریخ شروع باشد!');
            return;
        }

        const request = borrowManager.createBorrowRequest(this.currentUser.id, book.id, startDate, endDate);
        if (request) {
            console.log('درخواست امانت با موفقیت ثبت شد!');
        } else {
            console.log('خطا در ثبت درخواست!');
        }
    }

    showBorrowRequests(borrowManager) {
        console.log('\n--- درخواست‌های امانت شما ---');
        const requests = borrowManager.getStudentRequests(this.currentUser.id);
        
        if (requests.length === 0) {
            console.log('هیچ درخواست امانتی ندارید.');
            return;
        }

        requests.forEach((request, index) => {
            console.log(`${index + 1}. کتاب: ${request.bookTitle}`);
            console.log(`   تاریخ شروع: ${request.startDate}`);
            console.log(`   تاریخ پایان: ${request.endDate}`);
            console.log(`   وضعیت: ${request.status}`);
            console.log('');
        });
    }

    getStudentById(id) {
        return this.students.find(s => s.id === id);
    }

    getAllStudents() {
        return this.students;
    }

    getActiveStudents() {
        return this.students.filter(s => s.isActive);
    }

    toggleStudentStatus(studentId) {
        const student = this.students.find(s => s.id === studentId);
        if (student) {
            student.isActive = !student.isActive;
            return student.isActive;
        }
        return null;
    }
}

module.exports = StudentManager;

File: src/main/java/com/university/library/modules/StudentManager.java
package com.university.library.modules;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.function.Supplier;

public class StudentManager {
    private final List<Student> students;
    private int nextId;
    private Scanner scanner;
    private Supplier<Object> currentUserSupplier;

    public StudentManager() {
        this.students = new ArrayList<>();
        this.nextId = 1;
    }

    public void setInput(Scanner scanner, Supplier<Object> currentUserSupplier) {
        this.scanner = scanner;
        this.currentUserSupplier = currentUserSupplier;
    }

    public Student register() {
        System.out.println("\n--- Student Registration ---");
        System.out.print("Username: ");
        String username = scanner.nextLine().trim();
        if (students.stream().anyMatch(s -> s.username.equals(username))) {
            System.out.println("This username is already taken!");
            return null;
        }
        System.out.print("Password: ");
        String password = scanner.nextLine();
        System.out.print("Confirm password: ");
        String confirm = scanner.nextLine();
        if (!password.equals(confirm)) {
            System.out.println("Passwords do not match!");
            return null;
        }
        Student s = new Student(nextId++, username, password, true, LocalDate.now());
        students.add(s);
        System.out.println("Registration completed successfully!");
        return s;
    }

    public Student login(String username, String password) {
        return students.stream()
                .filter(s -> s.username.equals(username) && s.password.equals(password) && s.isActive)
                .findFirst().orElse(null);
    }

    public void searchBooks(BookManager bookManager) {
        System.out.println("\n--- Book Search ---");
        System.out.println("1. Search by title");
        System.out.println("2. Search by year");
        System.out.println("3. Search by author");
        System.out.println("4. Combined search");
        System.out.print("Choose: ");
        String choice = scanner.nextLine().trim();
        List<BookManager.Book> results = new ArrayList<>();
        switch (choice) {
            case "1" -> {
                System.out.print("Title: ");
                String t = scanner.nextLine();
                results = bookManager.searchByTitle(t);
            }
            case "2" -> {
                System.out.print("Year: ");
                String y = scanner.nextLine();
                results = bookManager.searchByYear(y);
            }
            case "3" -> {
                System.out.print("Author: ");
                String a = scanner.nextLine();
                results = bookManager.searchByAuthor(a);
            }
            case "4" -> {
                System.out.print("Title (optional): ");
                String t = scanner.nextLine();
                System.out.print("Year (optional): ");
                String y = scanner.nextLine();
                System.out.print("Author (optional): ");
                String a = scanner.nextLine();
                results = bookManager.combinedSearch(t, y, a);
            }
            default -> {
                System.out.println("Invalid choice!");
                return;
            }
        }
        if (results.isEmpty()) {
            System.out.println("No books found!");
        } else {
            System.out.println("\nResults (" + results.size() + " book(s)):");
            for (int i = 0; i < results.size(); i++) {
                var b = results.get(i);
                System.out.println((i + 1) + ". " + b.title() + " - " + b.author() + " (" + b.publicationYear() + ")");
                System.out.println("   Status: " + (b.isAvailable() ? "Available" : "Borrowed"));
                System.out.println("   ID: " + b.id());
                System.out.println();
            }
        }
    }

    public void requestBorrow(BookManager bookManager, BorrowManager borrowManager) {
        System.out.println("\n--- Create Borrow Request ---");
        System.out.print("Book ID: ");
        String idStr = scanner.nextLine();
        int id;
        try { id = Integer.parseInt(idStr.trim()); } catch (Exception ex) { System.out.println("Invalid ID!"); return; }
        var book = bookManager.getBookById(id);
        if (book == null) { System.out.println("Book not found!"); return; }
        if (!book.isAvailable()) { System.out.println("This book is currently borrowed!"); return; }

        System.out.print("Start date (YYYY-MM-DD): ");
        String startDate = scanner.nextLine().trim();
        System.out.print("End date (YYYY-MM-DD): ");
        String endDate = scanner.nextLine().trim();

        LocalDate start;
        LocalDate end;
        LocalDate today = LocalDate.now();
        try { start = LocalDate.parse(startDate); end = LocalDate.parse(endDate); } catch (Exception ex) {
            System.out.println("Invalid date format!");
            return;
        }
        if (start.isBefore(today)) { System.out.println("Start date cannot be before today!"); return; }
        if (!end.isAfter(start)) { System.out.println("End date must be after start date!"); return; }

        var cu = currentUserSupplier.get();
        if (!(cu instanceof Student s)) { System.out.println("Not authenticated as student."); return; }
        var req = borrowManager.createBorrowRequest(s.id, book.id(), startDate, endDate);
        if (req != null) {
            System.out.println("Borrow request submitted successfully!");
        } else {
            System.out.println("Failed to submit request!");
        }
    }

    public void showBorrowRequests(BorrowManager borrowManager) {
        System.out.println("\n--- Your Borrow Requests ---");
        var cu = currentUserSupplier.get();
        if (!(cu instanceof Student s)) { System.out.println("Not authenticated as student."); return; }
        var requests = borrowManager.getStudentRequests(s.id);
        if (requests.isEmpty()) {
            System.out.println("You have no borrow requests.");
            return;
        }
        for (int i = 0; i < requests.size(); i++) {
            var r = requests.get(i);
            System.out.println((i + 1) + ". Book: " + r.bookTitle());
            System.out.println("   Start date: " + r.startDate());
            System.out.println("   End date: " + r.endDate());
            System.out.println("   Status: " + r.status());
            System.out.println();
        }
    }

    public Student getStudentById(int id) { return students.stream().filter(s -> s.id == id).findFirst().orElse(null); }
    public List<Student> getAllStudents() { return students; }
    public List<Student> getActiveStudents() { return students.stream().filter(s -> s.isActive).toList(); }

    public Boolean toggleStudentStatus(int studentId) {
        Student s = getStudentById(studentId);
        if (s == null) return null;
        s.isActive = !s.isActive;
        return s.isActive;
    }

    public String getCurrentUsername(Object student) { return student instanceof Student s ? s.username : ""; }

    public record Student(int id, String username, String password, boolean isActive, LocalDate registrationDate) {}
}


